File: sketch/src/main/java/me/xiaopan/sketch/cache/LruBitmapPool.java
package me.xiaopan.sketch.cache;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Build;
import android.text.format.Formatter;
import android.util.Log;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import me.xiaopan.sketch.SLogType;
import me.xiaopan.sketch.SLog;
import me.xiaopan.sketch.cache.recycle.AttributeStrategy;
import me.xiaopan.sketch.cache.recycle.LruPoolStrategy;
import me.xiaopan.sketch.cache.recycle.SizeConfigStrategy;
import me.xiaopan.sketch.util.SketchUtils;

/**
 * An {@link BitmapPool} implementation that uses an
 * {@link me.xiaopan.sketch.cache.recycle.LruPoolStrategy} to bucket {@link Bitmap}s and then uses an LRU
 * eviction policy to evict {@link android.graphics.Bitmap}s from the least recently used bucket in order to keep
 * the pool below a given maximum size limit.
 */
public class LruBitmapPool implements BitmapPool {
    private static final Bitmap.Config DEFAULT_CONFIG = Bitmap.Config.ARGB_8888;

    private final LruPoolStrategy strategy;
    private final Set<Bitmap.Config> allowedConfigs;
    private final int initialMaxSize;
    private final BitmapTracker tracker;

    protected String logName = "LruBitmapPool";
    private int maxSize;
    private int currentSize;
    private int hits;
    private int misses;
    private int puts;
    private int evictions;

    private Context context;
    private boolean closed;
    private boolean disabled;

    // Exposed for testing only.
    LruBitmapPool(Context context, int maxSize, LruPoolStrategy strategy, Set<Bitmap.Config> allowedConfigs) {
        context = context.getApplicationContext();
        this.context = context;
        this.initialMaxSize = maxSize;
        this.maxSize = maxSize;
        this.strategy = strategy;
        this.allowedConfigs = allowedConfigs;
        this.tracker = new NullBitmapTracker();
    }

    /**
     * Constructor for LruBitmapPool.
     *
     * @param maxSize The initial maximum size of the pool in bytes.
     */
    public LruBitmapPool(Context context, int maxSize) {
        this(context, maxSize, getDefaultStrategy(), getDefaultAllowedConfigs());
    }

    /**
     * Constructor for LruBitmapPool.
     *
     * @param maxSize        The initial maximum size of the pool in bytes.
     * @param allowedConfigs A white listed set of {@link android.graphics.Bitmap.Config} that are allowed to be put
     *                       into the pool. Configs not in the allowed set will be rejected.
     */
    @SuppressWarnings("unused")
    public LruBitmapPool(Context context, int maxSize, Set<Bitmap.Config> allowedConfigs) {
        this(context, maxSize, getDefaultStrategy(), allowedConfigs);
    }

    private static LruPoolStrategy getDefaultStrategy() {
        final LruPoolStrategy strategy;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            strategy = new SizeConfigStrategy();
        } else {
            strategy = new AttributeStrategy();
        }
        return strategy;
    }

    private static Set<Bitmap.Config> getDefaultAllowedConfigs() {
        Set<Bitmap.Config> configs = new HashSet<Bitmap.Config>();
        configs.addAll(Arrays.asList(Bitmap.Config.values()));
        if (Build.VERSION.SDK_INT >= 19) {
            configs.add(null);
        }
        return Collections.unmodifiableSet(configs);
    }

    @Override
    public synchronized boolean put(Bitmap bitmap) {
        if (closed) {
            return false;
        }

        if (disabled) {
            SLog.w(SLogType.CACHE, logName, "Disabled. Unable put, bitmap=%s,%s",
                    strategy.logBitmap(bitmap), SketchUtils.toHexString(bitmap));
            return false;
        }

        if (bitmap == null) {
            throw new NullPointerException("Bitmap must not be null");
        }
        if (bitmap.isRecycled() || !bitmap.isMutable() || strategy.getSize(bitmap) > maxSize || !allowedConfigs.contains(bitmap.getConfig())) {
            SLog.w(SLogType.CACHE, logName, "Reject bitmap from pool, bitmap: %s, is recycled: %s, is mutable: %s, is allowed config: %s, %s",
                    strategy.logBitmap(bitmap), bitmap.isRecycled(), bitmap.isMutable(),
                    allowedConfigs.contains(bitmap.getConfig()), SketchUtils.toHexString(bitmap));
            return false;
        }

        final int size = strategy.getSize(bitmap);
        strategy.put(bitmap);
        tracker.add(bitmap);

        puts++;
        currentSize += size;

        SLog.v(SLogType.CACHE, logName, "Put bitmap in pool=%s,%s",
                strategy.logBitmap(bitmap), SketchUtils.toHexString(bitmap));
        dump();

        evict();
        return true;
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB_MR1)
    @Override
    public synchronized Bitmap getDirty(int width, int height, Bitmap.Config config) {
        if (closed) {
            return null;
        }

        if (disabled) {
            SLog.w(SLogType.CACHE, logName, "Disabled. Unable get, bitmap=%s,%s",
                    strategy.logBitmap(width, height, config));
            return null;
        }

        // Config will be null for non public config types, which can lead to transformations naively passing in
        // null as the requested config here. See issue #194.
        final Bitmap result = strategy.get(width, height, config != null ? config : DEFAULT_CONFIG);
        if (result == null) {
            SLog.w(SLogType.CACHE, logName, "Missing bitmap=%s", strategy.logBitmap(width, height, config));
            misses++;
        } else {
            SLog.i(SLogType.CACHE, logName, "Get bitmap=%s,%s", strategy.logBitmap(width, height, config), SketchUtils.toHexString(result));
            hits++;
            currentSize -= strategy.getSize(result);
            tracker.remove(result);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB_MR1) {
                result.setHasAlpha(true);
            }
        }
        dump();

        return result;
    }

    @Override
    public synchronized Bitmap get(int width, int height, Bitmap.Config config) {
        Bitmap result = getDirty(width, height, config);
        if (result != null) {
            // Bitmaps in the pool contain random data that in some cases must be cleared for an image to be rendered
            // correctly. we shouldn't force all consumers to independently erase the contents individually, so we do so
            // here. See issue #131.
            result.eraseColor(Color.TRANSPARENT);
        }

        return result;
    }

    @Override
    public Bitmap getOrMake(int width, int height, Bitmap.Config config) {
        Bitmap result = get(width, height, config);
        if (result == null) {
            result = Bitmap.createBitmap(width, height, config);
            if (SLogType.CACHE.isEnabled()) {
                StackTraceElement[] elements = new Exception().getStackTrace();
                StackTraceElement element = elements.length > 1 ? elements[1] : elements[0];
                SLog.i(SLogType.CACHE, String.format("Make bitmap. info:%dx%d,%s,%s - %s.%s:%d",
                        result.getWidth(), result.getHeight(), result.getConfig(), SketchUtils.toHexString(result),
                        element.getClassName(), element.getMethodName(), element.getLineNumber()));
            }
        }

        return result;
    }

    private void evict() {
        if (closed) {
            return;
        }

        trimToSize(maxSize);
    }

    @Override
    public int getMaxSize() {
        return maxSize;
    }

    @Override
    public int getSize() {
        return currentSize;
    }

    @Override
    public synchronized void setSizeMultiplier(float sizeMultiplier) {
        if (closed) {
            return;
        }

        maxSize = Math.round(initialMaxSize * sizeMultiplier);
        evict();
    }

    @Override
    public boolean isDisabled() {
        return disabled;
    }

    @Override
    public void setDisabled(boolean disabled) {
        this.disabled = disabled;
        if (disabled) {
            SLog.w(SLogType.CACHE, logName, "setDisabled. %s", true);
        } else {
            SLog.i(SLogType.CACHE, logName, "setDisabled. %s", false);
        }
    }

    @SuppressLint("InlinedApi")
    @Override
    public synchronized void trimMemory(int level) {
        long size = getSize();

        if (level >= android.content.ComponentCallbacks2.TRIM_MEMORY_MODERATE) {
            trimToSize(0);
        } else if (level >= android.content.ComponentCallbacks2.TRIM_MEMORY_BACKGROUND) {
            trimToSize(maxSize / 2);
        }

        if (SLogType.CACHE.isEnabled()) {
            String releasedSize = Formatter.formatFileSize(context, size - getSize());
            SLog.w(SLogType.CACHE, logName, "trimMemory. level=%s, released: %s",
                    SketchUtils.getTrimLevelName(level), releasedSize);
        }
    }

    @Override
    public synchronized void clear() {
        if (SLogType.CACHE.isEnabled()) {
            SLog.w(SLogType.CACHE, logName, "clear. before size %s", Formatter.formatFileSize(context, getSize()));
        }

        trimToSize(0);
    }

    @Override
    public synchronized boolean isClosed() {
        return closed;
    }

    @Override
    public synchronized void close() {
        if (closed) {
            return;
        }

        closed = true;
        trimToSize(0);
    }

    private synchronized void trimToSize(int size) {
        while (currentSize > size) {
            final Bitmap removed = strategy.removeLast();
            if (removed == null) {
                SLog.w(SLogType.CACHE, logName, "Size mismatch, resetting");
                dumpUnchecked();
                currentSize = 0;
                return;
            }

            SLog.e(SLogType.CACHE, logName, "Evicting bitmap=%s,%s",
                    strategy.logBitmap(removed), SketchUtils.toHexString(removed));
            tracker.remove(removed);
            currentSize -= strategy.getSize(removed);
            removed.recycle();
            evictions++;
            dump();
        }
    }

    private void dump() {
        if (Log.isLoggable(logName, Log.VERBOSE)) {
            dumpUnchecked();
        }
    }

    private void dumpUnchecked() {
        SLog.v(SLogType.CACHE, logName,
                "Hits=%d, misses=%d, puts=%d, evictions=%d, currentSize=%d, maxSize=%d, Strategy=%s",
                hits, misses, puts, evictions, currentSize, maxSize, strategy);
    }

    @Override
    public String getKey() {
        return String.format("%s(maxSize=%s)", logName, Formatter.formatFileSize(context, getMaxSize()));
    }

    private interface BitmapTracker {
        void add(Bitmap bitmap);

        void remove(Bitmap bitmap);
    }

    @SuppressWarnings("unused")
    // Only used for debugging
    private static class ThrowingBitmapTracker implements BitmapTracker {
        private final Set<Bitmap> bitmaps = Collections.synchronizedSet(new HashSet<Bitmap>());

        @Override
        public void add(Bitmap bitmap) {
            if (bitmaps.contains(bitmap)) {
                throw new IllegalStateException("Can't add already added bitmap: " + bitmap + " [" + bitmap.getWidth()
                        + "x" + bitmap.getHeight() + "]");
            }
            bitmaps.add(bitmap);
        }

        @Override
        public void remove(Bitmap bitmap) {
            if (!bitmaps.contains(bitmap)) {
                throw new IllegalStateException("Cannot remove bitmap not in tracker");
            }
            bitmaps.remove(bitmap);
        }
    }

    private static class NullBitmapTracker implements BitmapTracker {
        @Override
        public void add(Bitmap bitmap) {
            // Do nothing.
        }

        @Override
        public void remove(Bitmap bitmap) {
            // Do nothing.
        }
    }
}

File: sketch/src/main/java/com/github/panpf/sketch/cache/internal/LruBitmapPool.kt
/*
 * Copyright (C) 2022 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.cache.internal

import android.content.ComponentCallbacks2
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Build
import com.github.panpf.sketch.cache.BitmapPool
import com.github.panpf.sketch.decode.internal.logString
import com.github.panpf.sketch.util.Logger
import com.github.panpf.sketch.util.format
import com.github.panpf.sketch.util.formatFileSize
import com.github.panpf.sketch.util.getTrimLevelName
import com.github.panpf.sketch.util.pool.AttributeStrategy
import com.github.panpf.sketch.util.pool.LruPoolStrategy
import com.github.panpf.sketch.util.pool.SizeConfigStrategy
import java.util.concurrent.atomic.AtomicInteger
import kotlin.math.roundToInt

/**
 * Release the cached [Bitmap] reuse pool according to the least-used rule
 */
class LruBitmapPool constructor(
    override val maxSize: Long,
    val allowedConfigs: Set<Bitmap.Config?> =
        Bitmap.Config.values().run {
            if (Build.VERSION.SDK_INT >= 19) {
                listOf(null).plus(this).toSet()
            } else {
                this.toSet()
            }
        }
) : BitmapPool {

    companion object {
        private const val MODULE = "LruBitmapPool"
    }

    private var _size: Long = 0L
    private var hits = 0
    private var misses = 0
    private var puts = 0
    private var evictions = 0
    private val strategy: LruPoolStrategy = if (Build.VERSION.SDK_INT >= 19) {
        SizeConfigStrategy()
    } else {
        AttributeStrategy()
    }
    private val getCount = AtomicInteger()
    private val hitCount = AtomicInteger()

    override var logger: Logger? = null
    override val size: Long
        get() = _size

    override fun put(bitmap: Bitmap, caller: String?): Boolean {
        val bitmapKey = "${strategy.logBitmap(bitmap)}}"
        if (bitmap.isRecycled) {
            logger?.w(MODULE, "put. reject. Recycled. $caller. ${bitmap.logString}. $bitmapKey")
            return false
        }
        if (!bitmap.isMutable) {
            logger?.w(MODULE, "put. reject. Immutable. $caller. ${bitmap.logString}. $bitmapKey")
            return false
        }
        val bitmapSize = strategy.getSize(bitmap).toLong()
        if (bitmapSize > maxSize * 0.7f) {
            logger?.w(MODULE) {
                "put. reject. Too big ${bitmapSize.formatFileSize()}, maxSize ${maxSize.formatFileSize()}. $caller. ${bitmap.logString}. $bitmapKey"
            }
            return false
        }
        if (!allowedConfigs.contains(bitmap.config)) {
            logger?.w(MODULE) {
                "put. reject. Disallowed config ${bitmap.config}. $caller. ${bitmap.logString}. $bitmapKey"
            }
            return false
        }

        synchronized(this) {
            trimToSize(maxSize - bitmapSize, "$caller:putBefore")
            strategy.put(bitmap)
            puts++
            this._size += bitmapSize
            logger?.d(MODULE) {
                "put. successful. ${bitmap.logString}. size ${bitmapSize.formatFileSize()}, pool size ${size.formatFileSize()}. $caller. $bitmapKey"
            }
        }
        return true
    }

    override fun getDirty(width: Int, height: Int, config: Bitmap.Config): Bitmap? {
        // Config will be null for non public config types, which can lead to transformations naively passing in
        // null as the requested config here. See issue #194.
        return synchronized(this) {
            strategy[width, height, config].apply {
                val getCount = getCount.addAndGet(1)
                val hitCount = if (this != null) {
                    hitCount.addAndGet(1)
                } else {
                    hitCount.get()
                }
                if (getCount == Int.MAX_VALUE || hitCount == Int.MAX_VALUE) {
                    this@LruBitmapPool.getCount.set(0)
                    this@LruBitmapPool.hitCount.set(0)
                }
                if (this == null) {
                    misses++
                } else {
                    hits++
                    _size -= strategy.getSize(this)
                    this.setHasAlpha(true)
                }

                logger?.d(MODULE) {
                    val hitRatio = ((hitCount.toFloat() / getCount).format(2) * 100).roundToInt()
                    val bitmapKey = strategy.logBitmap(width, height, config)
                    if (this != null) {
                        "get. hit($hitRatio%). ${this.logString}. ${size.formatFileSize()}. $bitmapKey"
                    } else {
                        "get. miss($hitRatio%). ${size.formatFileSize()}. $bitmapKey"
                    }
                }
            }
        }
    }

    override fun get(width: Int, height: Int, config: Bitmap.Config): Bitmap? =
        getDirty(width, height, config)?.apply {
            eraseColor(Color.TRANSPARENT)
        }

    override fun getOrCreate(width: Int, height: Int, config: Bitmap.Config): Bitmap {
        return get(width, height, config) ?: Bitmap.createBitmap(width, height, config).apply {
            logger?.d(MODULE) {
                "getOrCreate. new. ${this.logString}. ${strategy.logBitmap(width, height, config)}"
            }
        }
    }

    override fun exist(width: Int, height: Int, config: Bitmap.Config): Boolean {
        return strategy.exist(width, height, config)
    }

    override fun trim(level: Int) {
        synchronized(this) {
            val oldSize = this.size
            if (level >= ComponentCallbacks2.TRIM_MEMORY_MODERATE) {
                trimToSize(0, "trim")
            } else if (level >= ComponentCallbacks2.TRIM_MEMORY_BACKGROUND) {
                trimToSize(maxSize / 2, "trim")
            }
            val releasedSize = (oldSize - size)
            logger?.w(MODULE) {
                "trim. level '${getTrimLevelName(level)}', released ${releasedSize.formatFileSize()}, size ${size.formatFileSize()}"
            }
        }
    }

    override fun clear() {
        synchronized(this) {
            val oldSize = size
            trimToSize(0, "clear")
            logger?.w(MODULE, "clear. cleared ${oldSize.formatFileSize()}")
        }
    }

    private fun trimToSize(size: Long, caller: String) {
        synchronized(this) {
            while (this.size > size) {
                val removed = strategy.removeLast()
                if (removed == null) {
                    this._size = 0
                } else {
                    this._size -= strategy.getSize(removed)
                    removed.recycle()
                    evictions++
                    logger?.d(MODULE) {
                        "trimToSize. Recycle bitmap. $caller. ${removed.logString}"
                    }
                }
            }
        }
    }

    override fun toString(): String {
        val strategy =
            if (strategy is SizeConfigStrategy) "SizeConfigStrategy" else "AttributeStrategy"
        val configs = allowedConfigs.joinToString(prefix = "[", postfix = "]", separator = ",")
        return "${MODULE}(maxSize=${maxSize.formatFileSize()},strategy=${strategy},allowedConfigs=${configs})"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is LruBitmapPool) return false
        if (maxSize != other.maxSize) return false
        if (allowedConfigs != other.allowedConfigs) return false
        return true
    }

    override fun hashCode(): Int {
        var result = maxSize.hashCode()
        result = 31 * result + allowedConfigs.hashCode()
        return result
    }
}